src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailService.kt

```diff
@@ -0,0 +1,120 @@
+package uk.gov.justice.digital.hmpps.manageusersapi.service.external
+
+import com.microsoft.applicationinsights.TelemetryClient
+import org.apache.commons.lang3.StringUtils
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.manageusersapi.adapter.auth.AuthApiService
+import uk.gov.justice.digital.hmpps.manageusersapi.adapter.auth.TokenByEmailTypeRequest
+import uk.gov.justice.digital.hmpps.manageusersapi.adapter.email.NotificationDetails
+import uk.gov.justice.digital.hmpps.manageusersapi.adapter.email.NotificationService
+import uk.gov.justice.digital.hmpps.manageusersapi.adapter.external.UserSearchApiService
+import uk.gov.justice.digital.hmpps.manageusersapi.resource.external.ExternalUserDetailsDto
+import uk.gov.justice.digital.hmpps.manageusersapi.service.EmailHelper
+
+@Service
+@Transactional(readOnly = true)
+class VerifyEmailService(
+  private val telemetryClient: TelemetryClient,
+  private val notificationService: NotificationService,
+  private val verifyEmailDomainService: VerifyEmailDomainService,
+  private val externalUserSearchApiService: UserSearchApiService,
+  private val authApiService: AuthApiService,
+  @Value("\${application.notify.verify.template}") private val notifyTemplateId: String
+) {
+
+  fun requestVerification(
+    userDetails: ExternalUserDetailsDto,
+    emailInput: String?,
+    url: String,
+  ): LinkEmailAndUsername {
+
+    val verifyLink =
+      url + authApiService.createTokenByEmailType(TokenByEmailTypeRequest(userDetails.username, EmailType.PRIMARY.name))
+
+    val email = EmailHelper.format(emailInput)
+    var usernameToUpdate = userDetails.username
+    validateEmailAddress(email, EmailType.PRIMARY)
+
+    if (userDetails.username.contains("@") && email!!.uppercase() != userDetails.username) {
+      usernameToUpdate = confirmUsernameForUpdate(email, userDetails.username)
+    }
+
+    val parameters: Map<String, Any> = mapOf(
+      "firstName" to userDetails.firstName,
+      "fullName" to "${userDetails.firstName} ${userDetails.lastName}",
+      "verifyLink" to verifyLink
+    )
+
+    notificationService.send(notifyTemplateId, parameters, "VerifyEmailRequest", NotificationDetails(userDetails.username, email!!))
+    return LinkEmailAndUsername(verifyLink, email, usernameToUpdate)
+  }
+
+  fun confirmUsernameForUpdate(newEmail: String, existingUsername: String): String {
```

Comment:
very minor - maybe rename to confirmUsernameValidForUpdate

Reply:
Done